### DISCUSSION TITLE
Is it possible to get a HTML+XPath feed from a Twitter page

### DISCUSSION THREAD

**Discussion by 38400000** [FreshRSS/FreshRSS#5498] [Q&A] [answered]
I have been experimenting with HTML + XPath (Web scraping), and have identified all the necessary queries to get the data needed for a complete feed of a Twitter page in FreshRSS.

I have tested them in the browser inspector and console to correctly return the content for all the different feed data types. Below are the queries.

```
XPath for: feed title
.//div[@data-testid="UserName"]

XPath for: finding news items (most important)
//article[@data-testid="tweet"]

XPath (relative to item) for: item title
.//div[@data-testid="tweetText"]

XPath (relative to item) for: item content
.//div[@class="news-content"]

XPath (relative to item) for: item link (URL)
.//div[@data-testid="User-Name"]//time/parent::*/@href

XPath (relative to item) for: item thumbnail
.//div[@data-testid="Tweet-User-Avatar"]//img/@src

XPath (relative to item) for: item author
.//div[@data-testid="User-Name"]/child::*[1]

XPath (relative to item) for: item date
.//div[@data-testid="User-Name"]//time/@datetime

XPath (relative to item) for: item unique ID
.//div[@data-testid="tweetText"]/@id
```

They were tested in the browser console with the following code:

```
//var xpath_query = '//div[@itemtype="http://schema.org/NewsArticle"]/@id'
//var xpath_query = '//time'
//var xpath_query = './/div[@data-testid="User-Name"]//time/parent::*/@href'
//var xpath_query = './/div[@data-testid="Tweet-User-Avatar"]//img/@src'
//var xpath_query = './/div[@data-testid="User-Name"]//time/@datetime'
var xpath_query = './/div[@data-testid="tweetText"]/@id'

var nodes = document.evaluate(xpath_query, document, null, XPathResult.ANY_TYPE, null);
var result = nodes.iterateNext();
while (result) {
  console.log(result.value);
  console.log(result.textContent);
  result = nodes.iterateNext();
}
```

However when i put it in FreshRSS, it fails to add the feed. The log file ("data\users\(user)\log.txt") shows the following error:

```
[Tue, 27 Jun 2023 18:14:31 +0200] [warning] --- Error fetching content: HTTP code 302: Maximum (4) redirects followed https://twitter.com/RetrogradeTom
[Tue, 27 Jun 2023 18:14:31 +0200] [warning] --- HTML+XPath Web scraping failed for [https://twitter.com/RetrogradeTom]
```

So my question is whether anyone has had any success getting HTML + XPath to work with a Twitter page, and if there is something i can do to get it working.

**Comment by Frenzie** [comment]
There's a max HTTP redirects setting per feed.

**Reply by 38400000** [reply]
Thanks.
I tried upping it to 50. But it seems like Twitter is sending it on an infinite series of redirects.

```
[Wed, 28 Jun 2023 20:16:12 +0200] [warning] --- Error fetching content: HTTP code 302: Maximum (50) redirects followed https://twitter.com/RetrogradeTom
[Wed, 28 Jun 2023 20:16:12 +0200] [warning] --- HTML+XPath Web scraping failed for [https://twitter.com/RetrogradeTom]
```

**Comment by Alkarex** [comment]
For fetching data as RSS from more complicated services, check projects such as https://github.com/RSS-Bridge/rss-bridge

**Reply by Frenzie** [reply]
Which I highly recommend, but for this specific use case it's of course also useless due to the API changes. Cf. https://github.com/RSS-Bridge/rss-bridge/issues/3462

**Reply by 38400000** [reply, marked as answer]
Thanks. Yes. I did use RSS-Bridge until Twitter locked anonymous access down.
Seems it is not possible at the current time, since it would need to use the user's own login session and simulate a browser session to scrape the webpage.

**Comment by gingerbeardman** [comment]
Not-logged-in viewing of timelines seems to be back, not sure to what extent. Maybe rate/view limited.

Long tweets have the "Show more" link which poses a problem to seeing full feed content.

Try this in a private browser tab https://twitter.com/darius

**Reply by Frenzie** [reply]
Not for me. Strange difference.

![image](https://github.com/FreshRSS/FreshRSS/assets/202757/1e37f2e7-9360-4645-8a38-7f578f8a6230)
![image](https://github.com/FreshRSS/FreshRSS/assets/202757/12289d7e-6aa6-433d-a43b-bf43318e8946)

**Reply by gingerbeardman** [reply]
Curious! Ah, well.

**Comment by Charlotte-br560** [comment]
Due to its dynamic nature and anti-scraping measures, directly scraping Twitter's HTML content is unlikely to be reliable. Instead, consider using Twitter's official APIs or third-party services like [Crawlbase](https://crawlbase.com/) to access structured data more reliably and sanctioned.
